### PR TITLE
Fix Bernoulli.log_prob boundary values at p=0 and p=1

### LINKED ISSUE #186825: Bernoulli(probs=0).log_prob(1) returns dtype-dependent finite values instead of -inf (Geometric special-cases this and tests it)
### 🐛 Describe the bug

`Bernoulli.log_prob` of an *impossible* outcome at the valid boundary parameters p=0 / p=1 returns a finite, dtype-dependent value instead of −inf:

```python
import torch

print(torch.distributions.Bernoulli(torch.tensor(0.0)).log_prob(torch.tensor(1.0)))
# tensor(-15.9424)   float32 — exp() ≈ 1.19e-7 for an event that cannot happen

print(torch.distributions.Bernoulli(torch.tensor(0.0, dtype=torch.float64)).log_prob(torch.tensor(1.0, dtype=torch.float64)))
# tensor(-36.0437)   float64 — same well-defined quantity, different answer
```

−15.9424 = `log(float32 eps)` and −36.0437 = `log(float64 eps)` — the result changes by ~20 nats with precision. Symmetric case: `Bernoulli(1.).log_prob(0.)`. Both p=0 and p=1 pass argument validation (the unit-interval constraint is inclusive) and value=1 is in the support, so validation does not catch this.

This contradicts the module's own tested semantics: `Geometric.log_prob` explicitly special-cases the boundary —

```python
probs = probs.clone(memory_format=torch.contiguous_format)
probs[(probs == 1) & (value == 0)] = 0
return value * (-probs).log1p() + self.probs.log()
```

— so `Geometric(1).log_prob(torch.tensor(1.0))` correctly returns `-inf`, and `test/distributions/test_distributions.py` **asserts exactly that** (`assertEqual(Geometric(1).log_prob(torch.tensor(1.0)), -inf)`). Exact −inf for impossible events at boundary p is tested, intended behavior in `torch.distributions`; Bernoulli just doesn't implement it.

Root cause: `Bernoulli.log_prob` always routes through logits — `probs_to_logits(probs, is_binary=True)` calls `clamp_probs`, which clamps p into `(eps, 1−eps)` before `binary_cross_entropy_with_logits`. The clamp is useful for gradient stability but leaks into the forward value.

Suggested fix: special-case exact-boundary probs in `log_prob` (mirroring Geometric), so the forward value is exact while gradients keep the clamped path. Possibly related (same clamp family): #72525.

For reference, `scipy.stats.bernoulli(0).logpmf(1)` returns `-inf`.

### Versions

torch 2.12.0, CPU (arm64, macOS); `Bernoulli.log_prob` unchanged on current main.

cc @fritzo @neerajprad @alicanb @nikitaved

### PR DESCRIPTION
**Pull request by karandhaodiyal28-hash**
Fixes #186825

## Motivation
`Bernoulli.log_prob` previously returned finite, dtype-dependent values for impossible events when initialized with exact boundary parameters `probs=0` or `probs=1`. For example, `Bernoulli(probs=0).log_prob(1)` produced `-15.94` in `float32` and `-36.04` in `float64` instead of exact `-inf`, because `probs_to_logits()` clamps probabilities into `[eps, 1-eps]`.

### Why this change is needed & cleaner:
1. **Mathematical Consistency:** Impossible events under boundary distributions must have probability 0, hence `log_prob == -inf`.
2. **Alignment across Distributions:** Matches the existing boundary behavior in sibling distributions like `Geometric`, where boundary probabilities explicitly handle edge cases.
3. **Reduces Unexpected Downstream Errors:** Returning finite log-probabilities for impossible events creates silent numerical drift in probabilistic inference and loss formulations.

### Potential Breakage & Mitigation:
- **Breakage:** Code relying on finite loss gradients for impossible boundary events under `probs=0` / `probs=1` may now receive `-inf` or NaNs during backprop if loss functions do not mask impossible events.
- **Mitigation:** Unaffected for non-boundary parameters ($0 < p < 1$). For boundary cases, using `logits` directly remains the recommended path for unclamped gradient flow. Updated docstrings explicitly highlight boundary semantics.

### CODE REVIEW

File: torch/distributions/bernoulli.py [hunk 1]
```diff
@@ -122,7 +122,25 @@ def log_prob(self, value):
...
+        # so those values are exact. The correction only runs when the
+        # distribution was constructed with `probs`: finite logits can never
+        # hit the exact boundary, and this keeps `probs` lazy for
+        # logits-constructed instances. See gh-186825.
+        if "probs" in self.__dict__:
+            probs, value = broadcast_all(self.probs, value)
+            boundary = (probs == 0) | (probs == 1)
+            impossible = boundary & (probs != value)
+            certain = boundary & (probs == value)
+            result = torch.where(
```
Comment (Skylion007) [inline torch/distributions/bernoulli.py]: Used masked_fill if you are going to do this.

**Comment by fritzo** [comment]
Please add Motivation section to PR description. Why is this a problem, what end use case is cleaner / less surprising by the change, what are the pros and cons? I expect this to break much code in the wild, please address breakage and docstring explanations you will add to mitigate breakage.

**Review by fritzo — CHANGES_REQUESTED** [review]
Motivation?

**Comment by karandhaodiyal28-hash** [comment]
Updated the PR description with a detailed Motivation & Breakage section, and added boundary regression tests covering float32/float64 in test_distributions.py.

**Comment by karandhaodiyal28-hash** [comment]
Hi @fritzo! Thanks for the review. The PR already includes a detailed Motivation section explaining why this change is needed — it covers mathematical consistency (impossible events must have log_prob == -inf), alignment with sibling distributions like Geometric, and the reduced downstream numerical errors. Would you mind taking another look? Happy to clarify or expand on any part. Thanks! 🙏

**Comment by karandhaodiyal28-hash** [comment]
@fritzo Hi! Gentle ping on this — it's been about 10 days since the changes-requested review.

The PR description now has a full **Motivation & Potential Breakage** section covering:
- Why boundary cases must return exact `-inf` / `0` (math consistency, alignment with `Geometric`, reduced downstream numerical drift)
- The con: downstream code that relied on finite gradients for impossible boundary events will now get `-inf` / NaN on backprop, plus the explicit mitigation: use `logits` directly for unclamped gradient flow
- A note in the updated docstring of `Bernoulli.log_prob` highlighting the boundary semantics

I also added boundary regression tests for both `float32` and `float64` in `test_distributions.py`.

If you'd prefer a different shape (e.g. a deprecation window, a stricter clamp, or restricting the fix to `logits`-based construction only), I'm happy to rework it — just let me know which direction you want.

Thanks! 🙏
